File: rust/cli/src/commands/encode.rs
//! `ball encode <source.rs>` — Rust -> Ball (issue #41).
use std::path::Path;

use clap::ValueEnum;

use crate::error::CliError;
use crate::output::{write_bytes, write_text};
use crate::panic_guard::catch_panic_message;
use crate::serialize::{program_to_binary, program_to_json};

/// Output format for `ball encode` (mirrors `dart/cli`'s `--format
/// json|binary`).
#[derive(Clone, Copy, Debug, ValueEnum)]
pub enum Format {
    /// Proto3 JSON, `@type`-enveloped (`.ball.json`) — human-readable, the
    /// default.
    Json,
    /// Raw protobuf binary (`.ball.bin`) — compact.
    Binary,
}

/// Read `path` as Rust source, encode it via `ball-lang-encoder`, and write the
/// resulting `ball.v1.Program` to `output` (or stdout) in `format`.
///
/// `ball_lang_encoder::encode` `panic!`s/`assert!`s on source it doesn't support
/// (no `fn main()`, an unsupported construct outside its documented Phase
/// 3a/3b scope — see `rust/encoder/src/lib.rs`'s module doc comment) —
/// [`catch_panic_message`] converts that into a [`CliError::Parse`] (exit
/// `2`) instead of aborting the process.
///
/// With `library` set (`ball encode --lib`, issue #491) the source is encoded
/// through `ball_lang_encoder::encode_library` instead: no `fn main()` is
/// required, and the emitted `Program` has an empty `entry_function`, so it is
/// deliberately not runnable (`ball check` reports "missing entry_function").
/// Every *other* documented-gap panic still surfaces as the same clean exit
/// `2` — the flag relaxes the entry-point requirement, nothing else.
///
/// With `crate_mode` set (`ball encode --crate`, issue #491) `path` is a crate
/// ROOT rather than one source file — the directory holding `Cargo.toml`, the
/// `src` directory, or the root `.rs` file. The crate's `mod` graph is walked
/// and every file encoded against one crate-wide symbol table, which is what
/// lets a `receiver.method(args)` whose method is declared in another file
/// resolve. The entry mode is then detected from the crate itself (a root
/// declaring `fn main` is runnable, a `lib.rs` without one is library mode);
/// `--lib` on top forces library mode. Reading the files is the walk's own job,
/// so a missing crate root surfaces through the same `catch_panic_message`
/// exit-`2` path as any other unsupported input.
pub fn encode(
    path: &Path,
    output: Option<&Path>,
    format: Format,
    library: bool,
    crate_mode: bool,
) -> Result<(), CliError> {
    let program = if crate_mode {
        let path = path.to_path_buf();
        catch_panic_message(move || {
            if library {
                ball_lang_encoder::encode_crate_library(&path)
            } else {
                ball_lang_encoder::encode_crate(&path)
            }
        })?
    } else {
        let source = std::fs::read_to_string(path)
            .map_err(|e| CliError::Io(format!("could not read {}: {e}", path.display())))?;
        catch_panic_message(move || {
            if library {
                ball_lang_encoder::encode_library(&source)
            } else {
                ball_lang_encoder::encode(&source)
            }
        })?
    };
    match format {
        Format::Json => write_text(output, &program_to_json(&program)?),
        Format::Binary => write_bytes(output, &program_to_binary(&program)),
    }
}

File: go/cli/encode.go
package cli

import (
	"flag"
	"io"
	"os"

	encoder "github.com/ball-lang/ball/go/encoder"
)

// cmdEncode implements `ball encode <source.go> [-lib] [-o out] [-format
// json|binary]`: read a Go source file, encode it into a ball.v1.Program via
// go/encoder, and write it to -o or stdout in the chosen format (JSON is the
// default, @type-enveloped; binary is the Any-wrapped canonical form).
//
// Without -lib the source must declare a `func main()` entry point. With -lib
// (issue #537) it need not: the file is encoded through encoder.EncodeLibrary,
// which produces a Program with an empty entry_function that `ball check`
// deliberately rejects as non-runnable. The flag is single-dash `-lib` because
// Go's flag package is single-dash and this CLI already uses -o/-format; the
// Rust CLI spells the same thing `--lib` and the C# CLI `--library` (see
// go/cli/AGENTS.md).
//
// The encoder fails loud: source it cannot represent (an unsupported construct
// outside its documented scope — see go/encoder/AGENTS.md) is returned as an
// error, surfaced here as a parseErr (exit 2), never a placeholder program.
func cmdEncode(args []string, w io.Writer) *cliError {
	const usage = "ball encode <source.go> [-lib] [-o out] [-format json|binary]"
	fs := flag.NewFlagSet("encode", flag.ContinueOnError)
	out := fs.String("o", "", "write the encoded program here instead of stdout")
	format := fs.String("format", "json", "output format: json | binary")
	lib := fs.Bool("lib", false,
		"encode a library file (no func main() required); the result has an empty entry_function and `ball check` rejects it as non-runnable")
	positionals, cerr := parseCommand(fs, "encode", usage, args)
	if cerr != nil {
		return cerr
	}
	if len(positionals) != 1 {
		return parseErr("encode: expected exactly one Go source path (usage: %s)", usage)
	}
	if *format != "json" && *format != "binary" {
		return parseErr("encode: unknown -format %q (want json or binary)", *format)
	}

	source, err := os.ReadFile(positionals[0])
	if err != nil {
		return ioErr("could not read %s: %v", positionals[0], err)
	}

	encode := encoder.Encode
	if *lib {
		encode = encoder.EncodeLibrary
	}
	prog, encErr := encode(string(source))
	if encErr != nil {
		return parseErr("encode: %v", encErr)
	}

	switch *format {
	case "binary":
		data, serr := programToBinary(prog)
		if serr != nil {
			return serr
		}
		return writeOut(w, *out, data)
	default:
		data, serr := programToJSON(prog)
		if serr != nil {
			return serr
		}
		return writeOut(w, *out, data)
	}
}
